test/unit/test_ranges.py

```diff
@@ -0,0 +1,310 @@
+# Copyright (c) 2015 Scality
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import contextlib
+import errno
+import gzip
+import httplib
+import json
+import os
+import os.path
+import pickle
+import shutil
+import sys
+import tempfile
+import time
+import unittest
+
+import eventlet
+
+import swift_scality_backend.server
+
+import swift.account.server
+import swift.container.server
+import swift.common.ring
+import swift.common.utils
+import swift.proxy.server
+
+try:
+    # Pre storage-policy compatibility
+    import swift.common.storage_policy
+except ImportError:
+    pass
+
+
+class TestRangeHeaders(unittest.TestCase):
+    """
+    Test outgoing range headers to sproxyd from the object server on a partial
+    get of an object. See
+    https://github.com/scality/ScalitySproxydSwift/issues/121 for more details.
+    """
+
+    def setUp(self):
+        if 'swift.common.storage_policy' in sys.modules:
+            # Clear out any previously set storage policies
+            swift.common.storage_policy.reload_storage_policies()
+
+        # Bind proxy, account, container, and object server to a free port
+        bind_address = ('127.0.0.1', 0)
+
+        # Directory holding account, container, and object rings
+        self.swift_dir = tempfile.mkdtemp()
+
+        # Setup mocked sproxyd server
+        self.sproxyd_socket = eventlet.listen(bind_address)
+        self.sproxyd = eventlet.spawn(eventlet.wsgi.server,
+                                      self.sproxyd_socket, self.sproxyd_mock)
+
+        self.sproxyd_path = '/proxy/chord'
+        conf = {
+            'swift_dir': self.swift_dir,
+            'devices': self.swift_dir,
+            'mount_check': 'false',
+            # Scality sproxyd config options
+            'sproxyd_host': '%s:%d' % self.sproxyd_socket.getsockname(),
+            'sproxyd_path': self.sproxyd_path,
+        }
+
+        # Hashes for the ring hashing algorithm when determining data placement
+        swift.common.utils.HASH_PATH_SUFFIX = 'foo'
+        swift.common.utils.HASH_PATH_PREFIX = 'bar'
+
+        # Setup account server
+        account_path = os.path.join(self.swift_dir, 'account.ring.gz')
```

Comment:
There's quite some duplication in the following 3 blocks. What about

```python
for ring_name, controller in [
        ('account', swift.account.server.AccountController),
        ('container', swift.container.server.ContainerController),
        ('object', swift_scality_backend.server.ObjectController),
        ]:
    path = os.path.join(self.swift_dir, '%s.ring.gz' % ring_name)
    socket = eventlet.listen(bind_address)
    setattr(self, '%s_socket' % ring_name, socket)
    port = socket.getsockname()[1]
    self.setup_ring(path, port=port)
    server = eventlet.spawn(eventlet.wsgi.server, socket, controller(conf))
    setattr(self, '%s_server' % ring_name, server)
```

Furthermore, instead of adding the `*_socket` and `*_server` values as attributes on `self`, you could keep a set of 'threads to be killed' and 'sockets to be closed' on the instance, and in the `tearDown` simply iterate over those sets and call the appropriate methods.